kyo-scheduler/jvm/src/main/scala/kyo/scheduler/Scheduler.scala

```diff
@@ -74,55 +68,54 @@ final class Scheduler(
-                val currentWorkers = this.currentWorkers
-                var position       = XSRandom.nextInt(currentWorkers)
-                var stride         = Math.min(currentWorkers, scheduleStride)
-                var minLoad        = Int.MaxValue
-                while (stride > 0 && minLoad != 0) {
-                    val candidate = workers(position)
-                    if (
-                        candidate != null &&
-                        (candidate ne submitter) &&
-                        candidate.checkAvailability(cycles)
-                    ) {
-                        val l = candidate.load()
-                        if (l < minLoad) {
-                            minLoad = l
-                            worker = candidate
-                        }
+        val nowMs          = clock.currentMillis()
+        var worker: Worker = null
+        if (submitter == null) {
+            worker = Worker.current()
+            if (worker != null && !worker.checkAvailability(nowMs))
+                worker = null
+        }
+        if (worker == null) {
+            val currentWorkers = this.currentWorkers
+            var position       = XSRandom.nextInt(currentWorkers)
+            var stride         = Math.min(currentWorkers, scheduleStride)
+            var minLoad        = Int.MaxValue
+            while (stride > 0 && minLoad != 0) {
+                val candidate = workers(position)
+                if (
+                    candidate != null &&
+                    (candidate ne submitter) &&
+                    candidate.checkAvailability(nowMs)
+                ) {
+                    val l = candidate.load()
+                    if (l < minLoad) {
+                        minLoad = l
+                        worker = candidate
                     }
-                    position += 1
-                    if (position == currentWorkers)
-                        position = 0
-                    stride -= 1
                 }
+                position += 1
+                if (position == currentWorkers)
+                    position = 0
+                stride -= 1
             }
-            while (worker == null)
-                worker = workers(XSRandom.nextInt(currentWorkers))
-            if (!worker.enqueue(cycles, task, force = tries >= scheduleTries))
-                loop(tries + 1)
         }
-        loop()
+        while (worker == null)
+            worker = workers(XSRandom.nextInt(currentWorkers))
+        worker.enqueue(task)
     }
 
     private def steal(thief: Worker): Task = {
-        val cycles         = this.cycles
+        val nowMs          = clock.currentMillis()
         val currentWorkers = this.currentWorkers
         var worker: Worker = null
         var maxLoad        = 1
-        var position       = 0
-        while (position < currentWorkers) {
+        var position       = XSRandom.nextInt(currentWorkers)
+        var stride         = Math.min(currentWorkers, stealStride)
```

Comment:
I've introduced a `stride` mechanism similar to `schedule` to reduce the amount of work during stealing. The previous impl could also become too expensive if there are too many blocked workers.